perfkitbenchmarker/providers/profitbricks/profitbricks_virtual_machine.py

```diff
@@ -0,0 +1,368 @@
+
+  Attributes:
+    cores: int. Number of CPU cores for a custom VM.
+    ram: int. Amount of RAM in MBs for a custom VM.
+  """
+
+  @classmethod
+  def _GetOptionDecoderConstructions(cls):
+    """Gets decoder classes and constructor args for each configurable option.
+
+    Returns:
+      dict. Maps option name string to a (ConfigOptionDecoder class, dict) pair.
+          The pair specifies a decoder class and its __init__() keyword
+          arguments to construct in order to decode the named option.
+    """
+    result = super(CustomMachineTypeSpec, cls)._GetOptionDecoderConstructions()
+    result.update({'cores': (option_decoders.IntDecoder, {'min': 1}),
+                   'ram': (option_decoders.IntDecoder, {'min': 1024})})
+    return result
+
+
+class MachineTypeDecoder(option_decoders.TypeVerifier):
+  """Decodes the machine_type option of a ProfitBricks VM config."""
+
+  def __init__(self, **kwargs):
+    super(MachineTypeDecoder, self).__init__((basestring, dict), **kwargs)
+
+  def Decode(self, value, component_full_name, flag_values):
+    """Decodes the machine_type option of a ProfitBricks VM config.
+
+    Args:
+      value: Either a string name of a PB machine type or a dict containing
+          'cpu' and 'memory' keys describing a custom VM.
+      component_full_name: string. Fully qualified name of the configurable
+          component containing the config option.
+      flag_values: flags.FlagValues. Runtime flag values to be propagated to
+          BaseSpec constructors.
+
+    Returns:
+      If value is a string, returns it unmodified. Otherwise, returns the
+      decoded CustomMachineTypeSpec.
+
+    Raises:
+      errors.Config.InvalidValue upon invalid input value.
+    """
+    super(MachineTypeDecoder, self).Decode(value, component_full_name,
+                                           flag_values)
+    if isinstance(value, basestring):
+      return value
+    return CustomMachineTypeSpec(self._GetOptionFullName(component_full_name),
+                                 flag_values=flag_values, **value)
+
+
+class ProfitBricksVmSpec(virtual_machine.BaseVmSpec):
+  """Object containing the information needed to create a
+  ProfitBricksVirtualMachine.
+
+  Attributes:
+    ram: None or int. RAM value in MB for custom ProfitBricks VM.
+    cores: None or int. CPU cores value for custom ProfitBricks VM.
+  """
+
+  CLOUD = providers.PROFITBRICKS
+
+  def __init__(self, *args, **kwargs):
+    super(ProfitBricksVmSpec, self).__init__(*args, **kwargs)
+    if isinstance(self.machine_type, CustomMachineTypeSpec):
+      logging.info('Using custom hardware configuration from config.')
+      self.cores = self.machine_type.cores
+      self.ram = self.machine_type.ram
+      self.machine_type = 'Custom (RAM: {}, Cores: {})'.format(self.ram,
+                                                               self.cores)
+    else:
+      try:
+          hardware = ast.literal_eval(self.machine_type)
```

Comment:
Ah - I think I see whats happening - the flag is getting passed through `yaml.load()` twice - once when the config is loaded and a second time in your `ApplyFlags` method.

The solution to this would be either to wrap the flag in quotes so it the load will work (it won't on dicts which is what it is in your example after the first load) like this:
```
ping:
  flags:
    machine_type: "{cores: 2, ram: 4096}"
```
Or you could put the machine type directly in the benchmark spec like this (since apply flags wont load the machine type a second time in this case):
```
ping:
  vm_groups:
    default:
      vm_spec:
        ProfitBricks:
          machine_type:
            cores: 2
            ram: 4096
```